### ISSUE TITLE
use different or configurable location for tmt cache

### ISSUE THREAD

**Issue by sopos** [teemtee/tmt#356]
I have quite limited free space in /var. Running just a few runs in a virtual machine consumes quite a lot of space. I have majority of the space available for my home.
Either use `~/.tmt` or `~/.cache/tmt` or `~/.local/tmp/tmt` instead of `/var/tmp/tmt`. This will also prevent collision of more users using tmt on the same machine where /var/tmp/tmt is kind of locked for the first user.

**Comment by pvalena** [comment]
Just as a workaround... does specifying the run ID like "`tmt run --id $my_new_run_path`" work for you?

I'd also like to ask, why would VM machines be stored in `/var/tmp/tmt`? In case of (rootless) containers, the runtime is in your home folder, as is, with `vagrant` f.e.. Or are you referring to `testcloud`? I'm not sure that one has user-support (if not, let's create issue for that).
Anyway, only the base images would be stored, and reused on subsequent runs. `finish` step run on every run is supposed to clean up the created runtime VM (IOW, no leftovers of VMs per runs). And, I imagine, if you intend to run them in parallel, you're dealing with memory issues anyway.

But you're right, we've discussed, previously, there should be some system (or user) -config ([docs](https://tmt.readthedocs.io/en/latest/stories/cli.html?highlight=config#config)). @psss what's the status of such config?

**Comment by sopos** [comment]
I think it is not convenient to prepare the patch on my own to avoid conflicts.
Also, what about running more instances under the same user, it this ok? I'm thinking of e.g. running Tiers on one Jenkins slave under the same user for more teams / Tiers levels (1, 2, 3), which is typical case for Tiers execution.

**Comment by t184256** [comment]
Let me also suggest /run/user/... and ~/.local/share ($XDG_RUNTIME_DIR and $XDG_DATA_HOME from [XDG Base Directory specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html)) as candidates for user-specific locations.

**Comment by pvalena** [comment]
> I think it is not convenient to prepare the patch on my own to avoid conflicts.
> Also, what about running more instances under the same user, it this ok? I'm thinking of e.g. running Tiers on one Jenkins slave under the same user for more teams / Tiers levels (1, 2, 3), which is typical case for Tiers execution.

Multiple runs/users *should* be fine by design (hopefully there're no race conditions). New folder is created for every `tmt run` (the --id part), so every run gets one and shouldn't touch anything else. I cannot speak for the provisioning though, as I'm not familiar with the implementation, but every provisioner should IMO support that.

_Note: for multiple users, I think you currently need to change the folder permissions, manually. This should be an RFE, @psss @thrix,  right?_

**Comment by thrix** [comment]
As discussed on the hacking session, the problem with the default directory for runs there should be a configuration knob for tmt, so people can easily change the default.

I personally like the idea of moving the default to a user specific directly, seems more reasonable default, given the fact more people can use systrems :)

> @sopos 
> I think it is not convenient to prepare the path on my own to avoid conflicts.

This is not feasible for the user, but I believe for each CI/testing system it is ia MUST :) In CI we generate ids ourselves, I would recommend each CI system to do that, not relying on any tool defaults.

> Also, what about running more instances under the same user, it this ok?

@psss is the code generating the run IDs safe in this manner?

> I'm thinking of e.g. running Tiers on one Jenkins slave under the same user for more teams / Tiers levels (1, 2, 3), which is typical case for Tiers execution.

You should change the run ID to the Jenkins workspace, so things like artifacts archiving work :) You should not rely on any default folders here. Jenkins runs should generate files in PWD, that is the common usage ...

**Comment by sopos** [comment]
> You should change the run ID to the Jenkins workspace, so things like artifacts archiving work :) You should not rely on any default folders here. Jenkins runs should generate files in PWD, that is the common usage ...

This completely makes sense. So for the Jenkins jobs it would be something, like `tmt run --id $PWD ...`

**Comment by jscotka** [comment]
There is also relevant issue inside cockpit.

When running cockpit tests (it dynamically downloads testing images for KVM) causing that copying data to temporary location is slow, and tooks big amount of space.
for for me there could help, to just copy git tracked files or do not copy anything and just run inside current location.

at least would be nice to warn, that it will copy non trivial amount of data to temporary location and possibly remove old logs(dirs) when doing that.

**Comment by psss** [comment]
If user feels safe, we could also support using symlinks instead of copying when `provision --how local` is enabled.

**Comment by jscotka** [comment]
symlinks to souces sounds interesting and good enough for me. but understand, that use it as default option could cause troubles with some interferces. (ideally it could do Copy on Write :-)

**Comment by psss** [comment]
Yes, definitely, this would be enabled only upon a user request (e.g. if user feels safe).

**Comment by psss** [comment]
On today's hacking session the following approach was proposed:

 * test runner: use `~/.cache/tmt` directory (under the home directory of the user who is executing `tmt`)
 * guest: use `~/.cache/tmt` directory (under the target user account which used for testing)

Does this make sense to others as well? Any concerns or objections?

**Comment by pholica** [comment]
Hello, I believe I've hit this issue as well. TMT uses /var/tmp/tmt by default and additionally, even if I specify `TMT_WORKDIR_ROOT=/home/some_user/tmt`, it tries to store the VM image in `/var/tmp/tmt/testcloud/images` effectively ignoring the `TMT_WORKDIR_ROOT` and provides misleading error message when it ran out of space:
```
Failed to prepare the image. Check the '/var/tmp/tmt/testcloud/images' directory permissions.
The exception was caused by the previous exception:
Problem writing to /var/tmp/tmt/testcloud/images/some_image.qcow2. Are you in group testcloud?
```

The command I used was: `tmt ... run ... -a provision -h virtual -c system -i $(pwd)/some_image.qcow2 ...`

$ rpm -q tmt tmt-provision-virtual
tmt-1.20.0-1.fc37.noarch
tmt-provision-virtual-1.20.0-1.fc37.noarch

**Comment by psss** [comment]
Thanks for reporting this. The `testcloud` issue should be fixed by #1779.

**Comment by pholica** [comment]
Thanks :)

**Comment by psss** [comment]
The `TMT_WORKDIR_ROOT` variable or `tmt run --workdir-root PATH` can be used to select the desired workdir root.
